llvm/lib/Target/WebAssembly/WebAssemblyMemorySafety.cpp

```diff
@@ -227,69 +286,115 @@ bool WebAssemblyMemorySafety::runOnFunction(Function &F) {
       }
       if (auto *Call = dyn_cast<CallInst>(&I)) {
         auto *CalledFunction = Call->getCalledFunction();
+
         inferNonMandatoryLibFuncAttrs(CalledFunction->getParent(), CalledFunction->getName(), TLIAnalysis.getTLI(F));
-        auto Attr =
-            CalledFunction->getFnAttribute(Attribute::AttrKind::AllocKind);
+        
+        auto Attr = CalledFunction->getFnAttribute(Attribute::AllocKind);
         if (Attr.hasAttribute(Attribute::AllocKind)) {
-          if (isAllocKind(Attr, AllocFnKind::Alloc)) {
-            CallsToAllocFunctions.emplace_back(
-                std::pair(AllocFnKind::Alloc, Call));
-          } else if (isAllocKind(Attr, AllocFnKind::Realloc)) {
-            CallsToAllocFunctions.emplace_back(
-                std::pair(AllocFnKind::Realloc, Call));
-          } else if (isAllocKind(Attr, AllocFnKind::Free)) {
-            CallsToAllocFunctions.emplace_back(
-                std::pair(AllocFnKind::Free, Call));
+          // Flatten the AllocKind so we can check if we want to/can replace
+          // this with our "__wasm_memsafety_" wrappers.
+          auto FlattenedAllocData = flattenAllocKind(Attr, Call, TLI);
+
+          if (FlattenedAllocData.kind == FlattenedAllocKind::Unhandled) {
+            // It's better to abruptly terminate than to allow an unsafe
+            // allocation function, which invalidates all kinds of system
+            // state/assumptions.
+            llvm::report_fatal_error("Unhandled alloc kind encountered", false);
           }
-          // if (isAllocKind(Attr, AllocFnKind::Uninitialized)) {
-          //   errs() << "AllocKind Uninitialized\n";
-          // }
-          // if (isAllocKind(Attr, AllocFnKind::Zeroed)) {
-          //   errs() << "AllocKind Zeroed\n";
-          // }
-          // if (isAllocKind(Attr, AllocFnKind::Aligned)) {
-          //   errs() << "AllocKind Aligned\n";
-          // }
+
+          CallsToAllocFunctions.emplace_back(FlattenedAllocData, Call);
         }
       }
     }
   }
 
+  auto SafeAlignedAllocFn = F.getParent()->getOrInsertFunction(
+      "__wasm_memsafety_aligned_alloc",
+      FunctionType::get(PointerType::getInt8PtrTy(Ctx),
+                        {
+                            Type::getInt64Ty(Ctx),
+                            Type::getInt64Ty(Ctx),
+                        },
+                        false));
   auto SafeMallocFn = F.getParent()->getOrInsertFunction(
       "__wasm_memsafety_malloc",
       FunctionType::get(PointerType::getInt8PtrTy(Ctx),
                         {
                             Type::getInt64Ty(Ctx),
+                        },
+                        false));
```

Comment:
This is probably of interest to you. I added a new `aligned_alloc`, which handles the cases when the C code contains `aligned_alloc`. What might be a little confusing is that the rtlib also contains an `aligned_alloc_for_mte`, but this is just an internal helper that does the MTE-required alignment, not any C-code-defined alignment (though it will be called after a `aligned_alloc(3)` specific check). I think adding this makes it clear which C allocation functions we support.